beluga/include/beluga/motion/ackermann_drive_model.hpp

```diff
@@ -0,0 +1,288 @@
+                          params_.steering_noise_from_steering * controls.phi * controls.phi)};
+    const auto steering_angle_params = steering_angle_distribution.param();
+
+    // Additional orientation noise (gamma_hat) using rotation parameters
+    const auto gamma_distribution = std::normal_distribution<double>{
+        0.0,  // zero mean
+        std::sqrt(
+            params_.orientation_noise_from_velocity * controls.v * controls.v +
+            params_.orientation_noise_from_steering * controls.phi * controls.phi)};
+    const auto gamma_params = gamma_distribution.param();
+
+    return [=](const auto& state, auto& gen) {
+      static thread_local auto distribution = std::normal_distribution<double>{};
+
+      // Sample noisy velocities
+      const auto v_hat = distribution(gen, linear_velocity_params);
+      const auto phi_hat = distribution(gen, steering_angle_params);
+      const auto gamma_hat = distribution(gen, gamma_params);
+
+      const auto omega_hat = v_hat * std::tan(phi_hat) / params_.wheelbase;
+      // Apply velocity motion model
+      return apply_velocity_motion(state, v_hat, omega_hat, gamma_hat, delta_time);
+    };
+  }
+
+  /// Calculate linear and angular velocities from two poses and delta time
+  AckermannControls calculate_velocities(
+      const Sophus::SE2d& pose,
+      const Sophus::SE2d& previous_pose,
+      std::chrono::duration<double> delta_time) const {
+    const double delta_t_sec = delta_time.count();
+
+    // Euclidean distance (chord length between poses)
+    const auto translation = pose.translation() - previous_pose.translation();
+    const double chord_distance = translation.norm();
+
+    const auto relative_transform = previous_pose.inverse() * pose;
+    // Angular velocity from orientation change
+    const auto angular_change = relative_transform.so2();
+    const double angle_change = angular_change.log();
+    const double angular_velocity = angle_change / delta_t_sec;
+
+    // Determine direction sign (forward/backward motion)
+    const double dx = relative_transform.translation().x();
+    const double sign = (dx >= 0.0) ? 1.0 : -1.0;
+
+    // Linear velocity calculation
+    double linear_velocity = 0.0;
+    double steering_angle = 0.0;
+    if (std::abs(angle_change) > small_angle_threshold) {
+      // Circular motion: calculate radius from chord and angle
+      // For an arc: chord = 2r·sin(θ/2), therefore r = chord / (2·sin(θ/2))
+      const double radius = chord_distance / (2.0 * std::sin(std::abs(angle_change) / 2.0));
+
+      // Arc length: s = r · θ
+      const double arc_distance = radius * std::abs(angle_change);
+
+      // Linear velocity with direction sign
+      linear_velocity = sign * arc_distance / delta_t_sec;
+      if (std::abs(linear_velocity) > small_angle_threshold) {
+        const double ratio = params_.wheelbase * angular_velocity / linear_velocity;
+        steering_angle = std::atan(ratio);
+      }
```

Comment:
I need to think this, the fact that we need to guard this is flagging me that there's an problem with how we are calculating motion.